### PR TITLE
Advanced Progress Bar

### PR DESCRIPTION
**Pull request by alanwguo**
This progress bar automatically shows progress by groupings. 

Things that belong to the same parent are all put in a group.
If a group has multiple children with the same name, those are merged together into a virtual group.

These virtual groups have different visual treatment because a virtual group should not add an additional level of nesting. 

Screenshot 2023-01-27 at 7 59 23 PM

TODO: Figure out bug where task associated with parent_task_id is missing. @rickyyx 

Signed-off-by: Alan Guo <aguo@anyscale.com>

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/experimental/state/common.py
Comment (alanwguo) [inline python/ray/experimental/state/common.py:705]: Is there a better way to determine if a task_id refers to the driver?

**Comment by rkooo567** [comment]
Is this ready to review?

File: python/ray/experimental/state/common.py
Comment (rickyyx) [reply python/ray/experimental/state/common.py:705]: theoretically `TaskState.type` should indicate if it's a driver task. But we are currently filtering out drivers by default.

I am working on a PR to change that.

**Comment by alanwguo** [comment]
> Is this ready to review?

Feel free to start reviewing. I don't expect major changes in behavior, just some additional small fixes.

File: python/ray/experimental/state/common.py
Comment (rkooo567) [reply python/ray/experimental/state/common.py:705]: Hmm not sure if it is also reliable. cc @rickyyx are we planning to have a PR to fix this? 

Maybe we can always include the driver task from the API, but we can filter it out when we print?

**Comment by rkooo567** [comment]
> the tasks that were created from different parents will be in different groups.

Does this mean if we have something like this;

1000 tasks that have child tasks.

we will have 1000 rows?

**Review by rkooo567 — COMMENTED** [review]
Given my question ^ is yes, I think advanced progress bar by lineage is not very comprehensive (whereas if we can build by data lineage, the live DAG bar, that could be useful). IIUC, this means if you have 1000 Trainer actors, executing 2 different types of tasks, it will create 1000 rows (Trainer_1, Trainer_2, ...), each row will contain 2 rows. I rather would like to see 1 row (Trainer: 1000) that has two sub rows (funcA, funcB). 

I think the current implementation could be a part of ray get tasks (that show the summary of child summary), but not the best for the advanced progress bar. Wdyt? 

This also makes it difficult to build source side summary (since we need to track all task ids to implement this)

**Comment by rickyyx** [comment]
From the implementation - it seems to me even if tasks have different parents, as long as their parents (or ancestors) have the same names, they will be merged together. Or maybe I am misunderstanding something - @alanwguo maybe could help clarify.

**Comment by ericl** [comment]
We shouldn't merge things. Aggregation should only ever be at the leaves, not in the middle of the tree.

To understand why--- in the 1k trainer case, you'd be unable to debug anything useful if the trainers all got mashed together. With them separated you can easily drill down into the progress of any individual trial.

**Comment by rkooo567** [comment]
@ericl I think that's what "lineage" is supposed to do. Although it didn't seem to work correctly when I tried with this script

```python3
import ray
ray.init()

@ray.remote
class Actor:
    def f(self):
        pass

    def g(self):
        pass

@ray.remote
def f():
    pass

ray.get([f.remote() for _ in range(30)])

@ray.remote
class B:
    def __init__(self):
        self.actors = [Actor.remote() for _ in range(10)]
        a = []
        b = []
        for actor in self.actors:
            a.extend([actor.f.remote() for _ in range(10)])
            b.extend([actor.g.remote() for _ in range(10)])

        ray.get(a)
        ray.get(b)
    def ready(self):
        pass

b = B.remote()
ray.get(b.ready.remote())
```

It has 
```
f 20/20
B 1
    A 10
        A.f 100
        A.g 100
```

Seems like lineage + name works correctly. 
Screen Shot 2023-01-26 at 2 11 48 AM

But I found this doesn't work when I ran one of air benchmark (the RayTrainWorker is not in the summary)
Screen Shot 2023-01-26 at 2 47 39 AM

by lineage doesn't work at all
Screen Shot 2023-01-26 at 2 12 14 AM

- f is not aggregated although it is a leaf
- Actor's tasks are not shown up (only some odd id that I cannot find what this is)

**Comment by rkooo567** [comment]
Btw it makes sense the aggregating only leaf nodes is useful. One concern is that with this approach, we will have 1 row per each Worker. I feel like this is unnecessary because that's what table is for.

I wonder if it makes sense to consider both Actor + actor tasks as the same node? In this way, we can avoid having 1 row per each TrainWorker while we can still keep Trainer -> Worker relationship. (and if users need to dig deeper than that, they can search from tables).

That said, if we have 3 Trainers with 10 Worker, each runs f each

```
Driver
    Trainer, actor_id ABC
        Workers 10
            f 10
    Trainer, actor_id ABC
        Workers 10
            f 10
    Trainer, actor_id ABC
        Workers 10
            f 10
```

instead of 

```
Driver
    Trainer, actor_id ABC
        Worker 1
            f 1
         Worker 2
            f 1
         ....
         Worker 10
             f 1
    Trainer, actor_id ABC
        Worker 1
            f 1
         Worker 2
            f 1
         ....
         Worker 10
             f 1
    Trainer, actor_id ABC
        Worker 1
            f 1
         Worker 2
            f 1
         ....
         Worker 10
             f 1
```

**Comment by rickyyx** [comment]
How hard would it be to have the collapsed row always show aggregated of the tree, but the expanded view shows the aggregated of each child tree? (i.e. aggregation happens at every subtree)

Example: 
```
> f 20/20
v B 211 (expanded)
    > A 210 (collapsed)
```
(Clicking on A expands A)
```
> f 20/20
v B 211 (expanded)
    v  A 210 (notice the aggregated count still valid on the A's row)
           > A 21 (collapsed A's) 
           > A 21
           > A 21
           ...
```
(Clicking on one of A expands its' subtree)
```
> f 20/20
v B 211 (expanded)
    v  A 210 (notice the aggregated count still valid on the A's row)
           v A 21           
                > A.f 10
                > A.g 11
           > A 21 (collapsed A's) 
           > A 21
           ...
```

**Comment by alanwguo** [comment]
TO clarify a few things.

@ericl  There are two aggregation styles selectable by dropdown.
1. Lineage_and_name: This groups things by owner (parent_task_id for regular tasks, actor_id for actor_tasks). Tasks are merged together if they have the same name.
2. Lineage: This merges things only at the leaf node. It groups things by owner (parent_task_id for regular tasks, actor_id for actor_tasks)

@rkooo567, if you select your visualization to be "Lineage", you'll see that it looks like the output that Eric expects. I think there is two things that might be confusing right now.

1) actors are named by their id because we currently cannot easily get the actor name. This will be unblocked once ricky allows us to get dead actors from the state-api actors endpoint.
2) Although we merge leaves, we don't merge leaves when they are at the top-level. This was a judgement call that people would want to things split out at the top level but we can remove this special case.

@rkooo567 for the broken train workflow. It might be the same reason that tune workloads are broken. @rickyyx is looking into this where tasks that are created from a separate thread ends up with a new task_id. Those tasks are not returned in the task backend so then we drop the entire tree.

@rickyyx, your suggestion should be possible... but I would prefer to tackle that as a follow-up in ray 2.4

I'll update the issue description with all of these things.

**Comment by alanwguo** [comment]
Working on tests now

**Comment by alanwguo** [comment]
@ericl @rkooo567 @rickyyx @scottsun94 , I ended up implementing hybrid approach.

This is ready for review now!

**Comment by rkooo567** [comment]
A couple high level feedback;  (planning for detailed review tmrw)

1. When we expand I found the summary disappears. Can we just keep them?
2. Do we need any pagination? I found it pretty difficult to see when there are many tasks/actors
3. We need task/actor id next to each entry otherwise although we know what's failing or going wrong we cannot figure out what they actually are.

**Comment by rkooo567** [comment]
Screen Shot 2023-01-28 at 9 36 43 AM

Btw, I found unexpandable row that has 0/2. What does this mean? Does it mean it has been retried?

**Comment by alanwguo** [comment]
> A couple high level feedback; (planning for detailed review tmrw)
> 
> 1. When we expand I found the summary disappears. Can we just keep them?
> 2. I think we need pagination? I found it pretty difficult to see when there are many tasks/actors.
> 3. We need task/actor id next to each entry otherwise although we know what's failing or going wrong we cannot figure out what they actually are.
> 4. Not sure if we have a solution here, it's pretty difficult to see the topology when things are nested. For this one, let's do more dogfooding...

1. The summary is only removed when expanded for "merged siblings" groups. This is because if we were to keep the summary row around, we would have to show the children of that group as nested in some way. But I think nesting it is confusing because these tasks are the same level, not one level lower. Ex: 
```
@ray.remote
def foo():
  ray.get(a.remote())
  ray.get(b.remote())
  ray.get(b.remote())
  ray.get(c.remote())
```
progress bar would look like this:
```
foo
  a
  b (x2)
    b
    b
  c
```
a, b, b, and c should all be siblings, but if we indent the b's, it looks like the b's are a lower level than a and c.


2. Pagination would not work well with expansion and collapse. It's possible you may not be able to collapse a group because it's parent is in a previous page. I think expansion and collapsion is likely enough.

3. Instead of showing id, it's probably better to link to the task or actor table below. I'd like to tackle that as a follow-up after this PR. It's possible I can take a look before branch cut.

4. we made some improvements here by moving the name closer to the progress bar. The topology is mostly show through the indentation of the name. I think we can look into doin some design where we show lines that point to the parent. Something like this:
Screenshot 2023-01-28 at 3 32 57 PM

**Comment by rkooo567** [comment]
Let's move forward with the current design from this PR. But I do think we need better design for 1 & 2... I am not very convinced that the current UI is good enough (I already found it difficult to use & hard to understand with only 10s of actors/tasks). But we can see & decide it from the dogfooding. 

For 3, I think linking is a great idea (definitely better than embedding IDs). I believe link means you will show the filtered table when you click them? Also, I think we can also do it in a separate PR, and it is not a blocker.

**Comment by scottsun94** [comment]
Do you have some good testing scripts? I tried examples of ray air and tune but none of them worked out of box. Not sure if it's related to local ray build..

**Comment by scottsun94** [comment]
RE the 4 discussion points above:
1. I could see why it makes from both side. We need feedback from more people to validate it. It makes sense to keep it as is.
2. + 1 that pagination doesn't make sense in this kind of expand/collapse pattern. For many tasks/actors case, we should find other solutions, probably something like task graph.
3. Could we use tooltip instead? Just hover on the task and show the metadata that matters in the tooltip so that they don't have to jump to somewhere else?
4. Same as 2, maybe task graph is a better solution for seeing topology?

**Comment by rkooo567** [comment]
Yeah, let's further discuss after merging this PR + do more dogfooding.. We should be aligned with what's the core workflow and the purpose of the visualization to actually discuss the best design.

**Comment by rkooo567** [comment]
@scottsun94 I could just simply use air example from the doc without issues

File: python/ray/experimental/state/common.py
Comment (rkooo567) [inline python/ray/experimental/state/common.py:1023]: This line is a bit hard to understand. Can you comment what `group.children[0]` means here?

**Review by rkooo567 — APPROVED** [review]

**Comment by rkooo567** [comment]
I think we need better testing than what we have now. But I am okay with merging the PR first and add tests in a follow up PR

File: python/ray/experimental/state/common.py
Comment (alanwguo) [reply python/ray/experimental/state/common.py:738]: One issue with that approach is if the list_tasks endpoint doesn't include the driver, we will drop the whole tree...

So going > 10k tasks has a very big chance of losing a lot of data.

File: python/ray/experimental/state/common.py
Comment (alanwguo) [reply python/ray/experimental/state/common.py:939]: when I say actor name here i mean the "actor class name". Right now it's parsed from the func name of it's initializer.

**Comment by alanwguo** [comment]
> I think we need better testing than what we have now (looks like currently we just infer list of tasks from expected output and re-verify it). But I am okay with merging the PR first and add tests in a follow up PR

okay, i will add a few more tests after branch cut.

**Comment by rkooo567** [comment]
As a follow up, please write additional tests!

**Pull request merged**
